wcfsetup/install/files/lib/system/service/worker/Encryption.class.php

```diff
@@ -0,0 +1,194 @@
+        \assert($newPublicKey, "Failed to serialize public key");
+        $sharedSecret = Encryption::getSharedSecret($userJwk, $newJwk);
+
+        // Section 3.3
+        $ikm = Encryption::getIKM($serviceWorker, $sharedSecret, $userPublicKey, $newPublicKey);
+        // Section 3.4
+        $salt = \random_bytes(16);
+        $content = Encryption::createContext($userPublicKey, $newPublicKey, $serviceWorker->contentEncoding);
+
+        $cek = \hash_hkdf(
+            Encryption::HASH_ALGORITHM,
+            $ikm,
+            16,
+            Encryption::createInfo($serviceWorker->contentEncoding, $content, $serviceWorker->contentEncoding),
+            $salt
+        );
+        $nonce = \hash_hkdf(
+            Encryption::HASH_ALGORITHM,
+            $ikm,
+            12,
+            Encryption::createInfo('nonce', $content, $serviceWorker->contentEncoding),
+            $salt
+        );
+        // Section 4
+        $payload = Encryption::addPadding($payload, $serviceWorker->contentEncoding);
+
+        $tag = '';
+        $encryptedText = \openssl_encrypt(
+            $payload,
+            'aes-128-gcm',
+            $cek,
+            OPENSSL_RAW_DATA,
+            $nonce,
+            $tag
+        );
+
+        if ($serviceWorker->contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            $headers['Encryption'] = 'salt=' . Base64UrlSafe::encode($salt);
+            $headers['Crypto-Key'] = 'dh=' . Base64UrlSafe::encode($newPublicKey);
+        }
+        $record = $encryptedText . $tag;
+
+        return Encryption::getEncryptionContentCodingHeader(
+            $serviceWorker->contentEncoding,
+            \mb_strlen($record, '8bit'),
+            $salt,
+            $newPublicKey
+        ) . $record;
+    }
+
+    private static function getSharedSecret(JWK $publicKey, #[\SensitiveParameter] JWK $privateKey): string
+    {
+        $result = \openssl_pkey_derive(
+            ECKey::convertPublicKeyToPEM($publicKey),
+            ECKey::convertPrivateKeyToPEM($privateKey),
+            256
+        );
+        \assert($result);
+        return \str_pad($result, 32, "\x00", STR_PAD_LEFT);
+    }
+
+    private static function addPadding(#[\SensitiveParameter] string $payload, string $contentEncoding): string
+    {
+        $length = \mb_strlen($payload, '8bit');
+        $paddingLength = ServiceWorkerHandler::MAX_PAYLOAD_LENGTH - $length;
+
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return \str_pad($payload . "\x02", $paddingLength + $length, "\x00", STR_PAD_RIGHT);
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return \pack('n*', $paddingLength) . \str_pad($payload, $paddingLength + $length, "\x00", STR_PAD_LEFT);
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
```

Comment:
I can't help but wonder if making the two possible content encodings an enum and adding a padding method would make this more readable. Something like this. Also note the use of `str_repeat()`:

```php
enum Encoding {
  case Aes128Gcm;
  case AesGcm;

  public function pad(#[\SensitiveParameter] string $payload) {
    $length = \mb_strlen($payload, '8bit');
    $paddingLength = ServiceWorkerHandler::MAX_PAYLOAD_LENGTH - $length;
    $padding = \str_repeat("\x00", $paddingLength);

    return match ($this) {
      self::Aes128Gcm => $payload . "\x02" . $padding,
      self::AesGcm => \pack('n', $paddingLength) . $padding . $payload,
    };
  }
}